### PR TITLE
NEW: refactor for sample peds for implementation of Feature peds

### PR DESCRIPTION
**Pull request by cherman2**
This is a refactor of sample PEDS for new implementation of feature peds. This PR also includes the implementation of Feature peds

Feature peds tracks what percentage of samples have a specific feature from the donor to investigate what features are engrafting in recipients.

### CODE REVIEW

**Comment by cherman2** [comment]
This pr currently has the sample peds refactor implemented and tested. As well as draft feature peds code with no unit test nor hookup to the q2-interface

**Comment by cherman2** [comment]
Need to add Unit tests for feature-ped and ~~solve issue with taxonomic label being cut off~~ (This will be a new PR)

**Comment by lizgehret** [comment]
> Need to add Unit tests for feature-ped and solve issue with taxonomic label being cut off

@cherman2 I see that you've since added tests for feature-peds, but does the taxonomic label being cut off need to be addressed in this PR? Or are you going to open up a separate PR for that?

**Comment by cherman2** [comment]
@lizgehret, I talked with evan and we decided it made more sense as a separate PR

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
    _check_duplicate_subject_timepoint(subject_series, metadata,
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py:75]: What's the difference between 'recipients with feature' and 'all possible recipients with feature'? Is this related to the recipients' donor?

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]: Should this method be `_check_subjects_in_all_timepoints` with timepoint as plural?

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
def _check_duplicate_subject_timepoint(subject_series, metadata,
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]: Not sure if the juice is worth the squeeze, but it might be nice to keep all of your variable naming conventions the same (i.e. snake case between words vs. no word separation like: donor_df vs. donordf)

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
            'description': 'Proportional Engraftment of Donor Strains'
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
            'description': 'Subject IDs linking samples across time'
```
This seemed more clear to me, but feel free to take it or leave it 🤷‍♀️

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
            'title': "Feature ID",
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
            'description': 'Proportional Engraftment of Donor Strains'
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
            'title': "Feature ID",
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]: I'm interested in what @ebolyen thinks about this - but I'm wondering if you can just combine the two filtering methods for the time column - checking for existence, and then the type (ensuring numeric) before returning it from the `_check_for_time_column` helper method.

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
def _filter_associated_references(reference_series, metadata, time_column,
```
Also can take it or leave it on this suggestion, but it might be a bit more clear to call this one a `filter` helper instead of a `check` helper - since you aren't checking for something specific here, just pulling out a subset of `reference_series` and returning that with the associated metadata.

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
def _check_subjects_in_all_timepoints(subject_series, num_timepoints,
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
def _create_recipient_table(reference_series, time_metadata, tabledf):
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
        reference_series[reference_series.index.isin(time_metadata.index)]
```

File: q2_fmt/_visualizer.py
Comment (lizgehret) [inline q2_fmt/_visualizer.py]:
```suggestion
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
```

**Review by lizgehret — COMMENTED** [review]
Overall this seems reasonable! Some small comments inline, nothing major.

File: q2_fmt/_peds.py
Comment (cherman2) [reply q2_fmt/_peds.py:75]: Recipient with features is the number of recipients that ended up with the donor feature, while all possible recipients with feature is the amount of people who could have received the feature

File: q2_fmt/_peds.py
Comment (cherman2) [reply q2_fmt/_peds.py]: I changed this not through github because I wanted to change all the calls of it.

File: q2_fmt/_peds.py
Comment (cherman2) [reply q2_fmt/_peds.py]: I am open to merging them. I just kinda like that it was only checking one specific thing. Made more sense in my head

File: q2_fmt/_peds.py
Comment (cherman2) [reply q2_fmt/_peds.py]: Changed this outside of github so I could address when it was called too.

File: q2_fmt/_peds.py
Comment (cherman2) [reply q2_fmt/_peds.py]: Did this in commit: [caf9a30](https://github.com/qiime2/q2-fmt/pull/48/commits/caf9a301c8fc7445945ae5177236c4bf478996bb)

**Comment by cherman2** [comment]
@lizgehret, This is ready for second review

**Comment by lizgehret** [comment]
Okay, technical review lgtm @cherman2 - I'm going to pull this down locally to test and then this should be g2g!

**Comment by lizgehret** [comment]
Help text looks good, general file structure for output QZA looks good. Some questions/comments post-testing:

- Should any column be allowed under `--p-subject-column`?
- When using a numeric column under `--p-reference-column` (e.g. time_point is what I used, testing with your example data) the following error occurs:
Screen Shot 2023-05-01 at 3 03 49 PM
I tested this with `time_point` and `SubjectID`, which both produced the same error. When I tried using `Sample ID`, I got this (slightly different) error:
Screen Shot 2023-05-01 at 3 07 43 PM
Not sure what limitations are intended for this column, but it might be worth adding a guard that will raise an error if you don't want any numeric columns being used here. Additionally, I might add a more verbose description in the latter error message for what should be used as a reference column.
- I tested out using a non-numeric column in `--p-time-column`, which produced the following error message:
```
Donor must be numeric
```
I would include the column name in this error message, along with the actual column that was included in the command (just for clarity's sake). Maybe something like:
```
Non-numeric values found in `--p-time-column`. Please make sure the column selected contains only simple integer values. Column with non-numeric values that was selected: "Donor"
```

**Review by lizgehret — CHANGES_REQUESTED** [review]
Some comments above post-user review! Just FYI I used the usage example dataset to test - it might be worth using a more 'real-world' dataset to test on after this round just to see if the actual results (vs. just the general file structure, etc) make sense.

**Comment by cherman2** [comment]
Thanks @lizgehret for your review! 
This ended up being a lot more changes then I was expecting so I thought I would list out what I changed to address your comments. 

1. I added a check subject column name function to feature peds. It already existed but I wasn't using it in feature peds. 
2. I wrote, and added to FP and SP, a new function that checks that subject column is not numeric.
        a.  I fixed my usage example metadata because the subject values were numeric lol.
        b. I added a unit-test for testing if subject column is numeric.
3. I wrote, and added to FP and SP, a new function that checks that reference column is not numeric. 
        a. I added a unit-test for testing if reference column is numeric.
4. I refactored the function that tests if time column is numeric to match how I checked metadata types in the above new functions.
         a. I added a unit-test for testing if time column is not numeric
5. I added a check that makes sure that the references are in the feature table before making my donor dataframe. This addresses the vague error you got when you used time_point as the reference. 
          a. I added a test to make sure this catches and lists reference values that are not in the feature table.
6. I now check to see if time, reference or subject are in the index before I say that the column name does not exist. This address the error you got when you put Sample ID as your reference name. 
          a.  I added 3 new unit test to check that time column, reference column, and subject column are not the metadata                               index
7. I re-wrote the "`blank` wasn't found in the metadata" error so that it says what parameter was being checked. (e.g. instead of saying "there was an error finding Sample ID in the metadata". It now says "There was an error finding the provided `--p-reference-column`: `Sample ID` in the metadata'".
      a. I rewrote the  test_incorrect_group_column_name unit-test for this new error message 
      b. I rewrote the test_incorrect_reference_column_name unit-test for this new error message
      c. I rewrote the test_incorrect_subject_column_name unit-test for this new error message
8. I implemented your suggested rewrite to my time column is not numeric error.

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py:201]: Hey @cherman2, I think you re-use iterations of this helper method enough times that it would be worth a slight refactor (post-2023.5, since I know we're right up against a deadline on this). I'd re-write something like this:
```
def _check_column_type(column_properties, column, column_type):
    try:
        assert column_properties[column].type == column_type
    except AssertionError as e:
        raise AssertionError('Non-%s values found in `--p-%s-column`. Please make sure the column selected contains the correct MetadataColumn type. Column with non-%s values that was selected: %s' % (column_type, column, column_type, column)) from e
```
I'm not sure if the syntax of the AssertionError message is exactly correct, but this is mainly just to show an example for how I'd refactor all of these helpers.

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py:216]: You could also refactor this branch and the `subject_column` one to be another helper that checks if the given column is equal to the `metadata.index.name`.

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py:919]: With the helper method refactor mentioned above, you won't need all of these individual column/type tests either - you can just have one that confirms the behavior given a specific column and column type.

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py:996]: Similar to last comment, if you refactor the subject/reference column check in metadata to be a single helper method, you can pair this down to just one test instead of two.

**Review by lizgehret — APPROVED** [review]
Hey @cherman2 overall this looks good to merge! A few suggestions in-line for some small refactors post-2023.5 release, since I realize this is time sensitive and none of the suggestions impact your code's actual functionality.

Only thing that you can actually update before I merge is a small print statement that got left behind in your debugging process. I included that as a suggestion in my inline comments, so you can just commit that and this will be g2g. Thanks for all of your hard work on this!

**Pull request merged**
